From d786a79f51ab47b1bb086bf67e78a1f28b7b9964 Mon Sep 17 00:00:00 2001
From: Bart Van Assche <bvanassche () acm ! org>
Date: Fri, 19 May 2023 22:03:46 +0000
Subject: [PATCH 4/5] block: BFQ: Move an invariant check

Check bfqq->dispatched for each BFQ queue instead of checking it for an
invalid bfqq pointer.

Fixes: 3e49c1e4a615 ("block: BFQ: Add several invariant checks")
Signed-off-by: Bart Van Assche <bvanassche@acm.org>
---
 block/bfq-iosched.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 8d45130ae..abeb6e560 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5406,6 +5406,7 @@ void bfq_put_queue(struct bfq_queue *bfqq)
 
 	WARN_ON_ONCE(!list_empty(&bfqq->fifo));
 	WARN_ON_ONCE(!RB_EMPTY_ROOT(&bfqq->sort_list));
+	WARN_ON_ONCE(bfqq->dispatched);
 
 	kmem_cache_free(bfq_pool, bfqq);
 	bfqg_and_blkg_put(bfqg);
@@ -7151,7 +7152,6 @@ static void bfq_exit_queue(struct elevator_queue *e)
 	for (actuator = 0; actuator < bfqd->num_actuators; actuator++)
 		WARN_ON_ONCE(bfqd->rq_in_driver[actuator]);
 	WARN_ON_ONCE(bfqd->tot_rq_in_driver);
-	WARN_ON_ONCE(bfqq->dispatched);
 
 	hrtimer_cancel(&bfqd->idle_slice_timer);
 
-- 
2.40.1.445.gf85cd430b1

